# **The Contributors to Poor Food Security Project**

The Contributors to Poor Food Security Project uses the pandas package for data analysis with python. It serves as a data filtering notebook that makes US food security data easier to access and use programmatically. 

This notebook acts as a tutorial on how to take tabular data from a public county data set, and systematically analyzes the food environment index, limited access to healthy foods, and median household income in Florida counties.

## **Overview of Tutorial**

**You will use this tutorial to cover the following processes:**

1. Importing the pandas package 

2. Creating a dataframe  

3. Using functions to filter through the dataframe to analyze the contributors to poor food security among Florida counties 

4. Concatenating data

5. Creating a subset and exporting as a new .cvs file 

# **Importing Pandas** 

**Packages** 

Packages give you additional tools and functions that are not provided in base Python. Pythons includes multiple packages to begin with. The Anaconda distribution comes with the “Pandas” package already installed.  

Once you have installed a package, you can load it into your current Python session with the import function. If not loaded, these new functions will not be available.  

**Pandas** 

Like spreadsheets in Microsoft Excel, Pandas allows us to store our data in tabular, multi-dimensional objects (dataframes) with familiar features like rows, columns, and headers. This is useful because it makes management, manipulation, and cleaning of large datasets much easier than would be the case using Python's built-in data structures such as lists. Pandas also provides a wide range of useful tools for working with data once it has been stored and structured. 

To begin, import the pandas package using the following command: 

In [6]:
import numpy as np
import pandas as pd

Notice that we load pandas with the usual import pandas and an extra as pd statement. This allows us to call functions from pandas with pd.<function> instead of pandas.<function> for convenience. as pd is not necessary to load the package. 

**Note:** we also imported the numpy package, which is going to help pandas do some of its math.

# **Creating a DataFrame** 

## **Working Directories & Relative Paths**  

The next step is to download the csv file “CountyHealthData_2014-2015.csv” and save your copy in the same folder as this Jupyter Notebook.  

**Note:** Your csv file must be saved in the same working directory as your .ipynb notebook file that you will use.  

**Remember:** Jupyter Notebooks automatically set your working directory to the folder where the .ipynb is saved. You'll have to save the document at least once to set your directory, but once there you can use what's called relative file paths to access the files there.) 

If a file is located in your working directory, its relative path is just the name of the file!  

**Using the pd.read_csv( ) function** 

pd.read_csv reads the tabular data from a Comma Separated Values (csv) file into a dataframe object that we'll define as df. 

To create our dataframe object we'll define our object df by executing the pd.read_csv()function on our data file by inserting the relative file path into the parathenses.  

In [10]:
df=pd.read_csv("CountyHealthData_2014-2015 (3).csv")

**Note:** You will need to create your dataframe object each time you open python. To do this, press Enter and Ctrl to refresh your dataframe.

**Note:** You must write the exact title of your file in quotations in order for it to be read.

# **Filtering the Data**  

**Indexing**  

You will first want to manipulate the data to analyze the food environment indexes in Florida.

You will do this by selecting certain rows or columns from the larger dataframe. 

**Selecting the Rows** 

To analyze data from only Florida, you need to select the row numbers that correspond with the state "FL." These are rows 618 through 749. 

**Note:** When you select a range of rows, your return will give you data from the minimum row number through one lower than your maximum row number. 

For example, if you code your range as [618:749], your data will visibly return rows 618-748. 

You must make your maximum range number one row higher than you want in your table. 

Coding your range as [618:750] will visibly display rows 618-749. 

As shown above, these rows will be represented as numbers in square brackets. The first row in the dataset corresponding to "FL" is number 618. The last row in the dataset corresponding to "FL" is number 749. 

Nest these numbers inside square brackets with a colon in between.

**Selecting the Columns** 

In order to display the columns "State" and "Food environemnt index," you need to select the columns that are labeled as so. 

Start by adding the "State" column.

Nest "State" in its own set of square brackets and quotations before the brackets with your selected rows. 

**It should look like this:** 

In [18]:
df["State"][618:750] 

618    FL
619    FL
620    FL
621    FL
622    FL
       ..
745    FL
746    FL
747    FL
748    FL
749    FL
Name: State, Length: 132, dtype: object

Now that you have represented all of Florida's row in your table, you need to add the column "Food environment index."

Nest "Food environment index" in its own set of square brackets and quotations after "State," with a comma separating the two sets of quotations. Add another set of brackets around the previous column brackets.

**Note:** Your column name must have the same exact title as the column in your .csv in order for your table to be generated. This includes spaces and upper-case/lower-case letters. 

**It should look like this:**

In [19]:
df[["State", "Food environment index"]][618:750]

State  Food environment index
618    FL                   6.614
619    FL                   6.400
620    FL                   6.983
621    FL                   6.600
622    FL                   7.624
..    ...                     ...
745    FL                   8.000
746    FL                   7.770
747    FL                   7.300
748    FL                   5.651
749    FL                   5.300

[132 rows x 2 columns]

After filtering through the data you are able to see the food environment indexes for 132 rows labeled "FL" 

Why are there so many different rows labeled "FL"?

Each row represents a different county. However, you cannot see that in your table yet. 

# **Adding another column to describe your data**

After filtering through the data you are able to see the food environment indexes for 132 rows labeled "FL" 

Why are there so many different rows labeled "FL"?

Each row represents a different county. However, you cannot see that in your table yet. 

To add a column to label the different counties in Florida, you will add another set of quotations to your column bracket. The column name should be "County."

For reading purposes, nest the label "County" inbetween labels "State" and "Food environment index."

Don't forget to add another comma to separate each column name. 

**It should now look something like this:**

In [22]:
df[["State", "County", "Food environment index"]][618:750]

State             County  Food environment index
618    FL       Baker County                   6.614
619    FL       Baker County                   6.400
620    FL         Bay County                   6.983
621    FL         Bay County                   6.600
622    FL    Bradford County                   7.624
..    ...                ...                     ...
745    FL     Wakulla County                   8.000
746    FL      Walton County                   7.770
747    FL      Walton County                   7.300
748    FL  Washington County                   5.651
749    FL  Washington County                   5.300

[132 rows x 3 columns]

 **After filtering through the data you are able to see the food environment indexes for 132 counties in Florida.**

# **Adding Columns to further analyze**

The food environment index is a metric that measures the availability of healthy food within a relative distance. In order to ensure food security, people must have physical and economic access to sufficient amounts of nutrition.

You can further analyze the causes of Florida’s varying food environment indexes by comparing that data to data pertaining to limited accessibility to healthy food and income.

**You can do this by adding two more columns to your DataFrame.**  

Place a comma after “Food environment index” and write “Limited access to healthy foods”. 

Then, place another comma after “Limited access to healthy foods” and add “Median household income”. 

Don't forget to add commas between each column label. 

**Your new DataFrame code should look like this:**

In [27]:
df[["State", "County", "Food environment index", "Limited access to healthy foods", "Median household income"]][618:750]

State             County  Food environment index  \
618    FL       Baker County                   6.614   
619    FL       Baker County                   6.400   
620    FL         Bay County                   6.983   
621    FL         Bay County                   6.600   
622    FL    Bradford County                   7.624   
..    ...                ...                     ...   
745    FL     Wakulla County                   8.000   
746    FL      Walton County                   7.770   
747    FL      Walton County                   7.300   
748    FL  Washington County                   5.651   
749    FL  Washington County                   5.300   

     Limited access to healthy foods  Median household income  
618                            0.128                    45721  
619                            0.128                    46505  
620                            0.096                    45259  
621                            0.096                    45249  
622                            0.023                    38680  
..                               ...                      ...  
745                            0.003                    47430  
746                            0.072                    40239  
747                            0.072                    45357  
748                            0.194                    36560  
749                            0.194                    35671  

[132 rows x 5 columns]

**Now you have one big table that provides the name of each Florida county, its corresponding food environment index, access to healthy food, and median income.**    

## **Condensing the Data** 

The filtered data set that you just created includes 132 rows of data. 

Food environment indexes are given on a scale of 0-10. 

0 corresponds to the worst scoring, and 10 corresponds to the best scoring.  

To make this data set easier to analyze, let's condense it to counties in Florida that have a food environment index of 6 or lower.

When you look at the "Food Environment Index," for all the counties in Florida, you can find that there are 17 counties with a food environment index that is equal to or less than 6.  

**The corresponding index numbers to these counties are: 629, 647, 652, 653, 663, 661, 662, 663, 664, 665, 687, 693, 708, 709, 723, 748, and 749.** 

When choosing what rows to include in your table, we learned that we must provide a starting row, and an ending row. This will output a table that starts at your starting point, and includes every number until the ending point.  

However, since you are now using rows that are not in chronological order, we must make separate tables, and then combine them. 


**You will end up creating 9 separate tables.**

**Table 0**: Row 629

**Table 1**: Row 647

**Table 2**: Row 652 and 653

**Table 3**: Row 660, 661, 662, 663, 664, 665

**Table 4**: Row 687

**Table 5**: Row 693

**Table 6**: Row 708 and 709

**Table 7**: Row 723

**Table 8**: Row 748 and 749

You will start off by making your table consist of row 629.

This means your range will be [629:630] since row 630 will not be visible on your table.


**Your code for this condensed table should look like this:**


In [42]:
df[["State", "County", "Food environment index", "Limited access to healthy foods", "Median household income"]][629:630]

State          County  Food environment index  \
629    FL  Calhoun County                     5.9   

     Limited access to healthy foods  Median household income  
629                            0.132                    33977

You must identify each table you create.

Identify this table as "df0"

You will equal this identifier to your code. 

**It should look like this:**

In [43]:
df0= df[["State", "County", "Food environment index", "Limited access to healthy foods", "Median household income"]][629:630]


Now we will make a table for row 647.

This means your range will be [647:648] since row 648 will not be visible on your table.

**Your code for this condensed table should look like this:**

In [45]:
df[["State", "County", "Food environment index", "Limited access to healthy foods", "Median household income"]][647:648]

State           County  Food environment index  \
647    FL  Escambia County                     5.8   

     Limited access to healthy foods  Median household income  
647                             0.13                    42742

Identify this table as "df1"

**It should look like this:**

In [47]:
df1=df[["State", "County", "Food environment index", "Limited access to healthy foods", "Median household income"]][647:648]

Now we will make a table for rows 652 and 653.

This means your range will be [652:654] since row 654 will not be visible on your table.

**Your code for this condensed table should look like this:**

In [50]:
df[["State", "County", "Food environment index", "Limited access to healthy foods", "Median household income"]][652:654]

State          County  Food environment index  \
652    FL  Gadsden County                   5.706   
653    FL  Gadsden County                   5.500   

     Limited access to healthy foods  Median household income  
652                            0.058                    35854  
653                            0.058                    34348

Identify this table as "df2"

**It should look like this:**

In [69]:
df2=df[["State", "County", "Food environment index", "Limited access to healthy foods", "Median household income"]][652:654]

Now we will make a table for rows 660 through 665.

This means your range will be [660:666] since row 666 will not be visible on your table.

**Your code for this condensed table should look like this:**

In [52]:
df[["State", "County", "Food environment index", "Limited access to healthy foods", "Median household income"]][660:666]

State           County  Food environment index  \
660    FL  Hamilton County                   5.896   
661    FL  Hamilton County                   5.600   
662    FL    Hardee County                   5.642   
663    FL    Hardee County                   5.500   
664    FL    Hendry County                   5.660   
665    FL    Hendry County                   5.600   

     Limited access to healthy foods  Median household income  
660                            0.103                    32232  
661                            0.103                    33836  
662                            0.203                    34012  
663                            0.203                    34884  
664                            0.143                    36515  
665                            0.143                    37085

Identify this table as "df3"

**It should look like this:**

In [55]:
df3=df[["State", "County", "Food environment index", "Limited access to healthy foods", "Median household income"]][660:666]

Now we will make a table for row 687.

This means your range will be [687:688] since row 688 will not be visible on your table.

**Your code for this condensed table should look like this:**

In [56]:
df[["State", "County", "Food environment index", "Limited access to healthy foods", "Median household income"]][687:688]

State       County  Food environment index  \
687    FL  Leon County                     5.8   

     Limited access to healthy foods  Median household income  
687                            0.083                    46975

Identify this table as "df4"

**It should look like this:**

In [58]:
df4=df[["State", "County", "Food environment index", "Limited access to healthy foods", "Median household income"]][687:688]

Now we will make a table for row 693.

This means your range will be [693:694] since row 694 will not be visible on your table.

**Your code for this condensed table should look like this:**

In [60]:
df[["State", "County", "Food environment index", "Limited access to healthy foods", "Median household income"]][693:694]

State          County  Food environment index  \
693    FL  Madison County                     5.7   

     Limited access to healthy foods  Median household income  
693                            0.087                    30391

Identify this table as "df5"

**It should look like this:**

In [61]:
df5=df[["State", "County", "Food environment index", "Limited access to healthy foods", "Median household income"]][693:694]

Now we will make a table for rows 708 and 709.

This means your range will be [708:710] since row 710 will not be visible on your table.

**Your code for this condensed table should look like this:**

In [62]:
df[["State", "County", "Food environment index", "Limited access to healthy foods", "Median household income"]][708:710]

State             County  Food environment index  \
708    FL  Okeechobee County                   5.417   
709    FL  Okeechobee County                   5.100   

     Limited access to healthy foods  Median household income  
708                            0.215                    34742  
709                            0.215                    35954

Identify this table as "df6"

**It should look like this:**

In [63]:
df6=df[["State", "County", "Food environment index", "Limited access to healthy foods", "Median household income"]][708:710]

Now we will make a table for row 723.

This means your range will be [723:724] since row 724 will not be visible on your table.

**Your code for this condensed table should look like this:**

In [64]:
df[["State", "County", "Food environment index", "Limited access to healthy foods", "Median household income"]][723:724]

State         County  Food environment index  \
723    FL  Putnam County                     5.6   

     Limited access to healthy foods  Median household income  
723                            0.126                    32852

Identify this table as "df7"

**It should look like this:**

In [65]:
df7=df[["State", "County", "Food environment index", "Limited access to healthy foods", "Median household income"]][723:724]

Now we will make a table for rows 748 and 749.

This means your range will be [748:750] since row 750 will not be visible on your table.


**Your code for this condensed table should look like this:**


In [66]:
df[["State", "County", "Food environment index", "Limited access to healthy foods", "Median household income"]][748:750]

State             County  Food environment index  \
748    FL  Washington County                   5.651   
749    FL  Washington County                   5.300   

     Limited access to healthy foods  Median household income  
748                            0.194                    36560  
749                            0.194                    35671

Identify this table as "df8"

**It should look like this:**

In [67]:
df8=df[["State", "County", "Food environment index", "Limited access to healthy foods", "Median household income"]][748:750]

Now that we have created nine separate tables, we need to glue them together through a process called concatenating.  

# **Concatenating**

Concatenating data is most useful when we have to datasets that are basically identical in shape, and we simply need to 'glue' them together. 

It seems clear that these dataframea would be best combined by stacking them on top of eachother, or appending one to the other as additional rows or observations. 

**Panda's pd.concat lets us concatenate a list of dataframes into a single dataframe.**

We can do this by coding:
    pd.concat([df0,df1,df2,df3,df4,df5,df6,df7,df8],axis=0,ignore_index=True,sort=False)


**Notes:**

-axis=0 indicates that we want to add the dataframes together row-wise. What happens if we change to axis=1? 

-Pandas thinks about dimensions as rows and columns, in that order. axis=0 refers to rows, whereas axis=1 refers to columns. 

-ignore_index=True, resets the dataframe index to start at 0 and run to 3. Otherwise our row index would be 0 1 0 1, from the indices of the original two dataframes. 

-sort=False addresses a behavior for sorting in my version of python that causes an error when non-concatenation axis is not aligned. The value False tells it to ignore this sorting behavior 


In [70]:
pd.concat([df0,df1,df2,df3,df4,df5,df6,df7,df8],axis=0,ignore_index=True,sort=False)

State             County  Food environment index  \
0     FL     Calhoun County                   5.900   
1     FL    Escambia County                   5.800   
2     FL     Gadsden County                   5.706   
3     FL     Gadsden County                   5.500   
4     FL    Hamilton County                   5.896   
5     FL    Hamilton County                   5.600   
6     FL      Hardee County                   5.642   
7     FL      Hardee County                   5.500   
8     FL      Hendry County                   5.660   
9     FL      Hendry County                   5.600   
10    FL        Leon County                   5.800   
11    FL     Madison County                   5.700   
12    FL  Okeechobee County                   5.417   
13    FL  Okeechobee County                   5.100   
14    FL      Putnam County                   5.600   
15    FL  Washington County                   5.651   
16    FL  Washington County                   5.300   

    Limited access to healthy foods  Median household income  
0                             0.132                    33977  
1                             0.130                    42742  
2                             0.058                    35854  
3                             0.058                    34348  
4                             0.103                    32232  
5                             0.103                    33836  
6                             0.203                    34012  
7                             0.203                    34884  
8                             0.143                    36515  
9                             0.143                    37085  
10                            0.083                    46975  
11                            0.087                    30391  
12                            0.215                    34742  
13                            0.215                    35954  
14                            0.126                    32852  
15                            0.194                    36560  
16                            0.194                    35671

Now you have the access to healthy foods and median income of Florida counties that have a food environment index equal to or less than 6.

Through the data filtering process, you can see if the factors "limited access to healthy foods" and "median household income" are contributors to poor food environment indexes by comparing their numeric data.

**Now it's time to  name your table to make it ready for export.**

# **Exporting**

In order to export your table, you need to name it.

Name your table "Concat"

You will set "Concat" equal to your code of your concatenated table you made above. 


**It should look like this:**

In [87]:
Concat=pd.concat([df0,df1,df2,df3,df4,df5,df6,df7,df8],axis=0,ignore_index=True,sort=False)

### **Exporting to .csv file**

Now that you've finished manipulating your dataset for further analysis, you can export it as new .csv files, giving you a readymade and openly accessible output to share with the public on our GitHub repositories.

To do this you can use the method `.to_csv()` - adding the filename and extension within the parentheses at the end.

**It should look like this:**

In [86]:
Concat.to_csv("MergedData.csv", index=False)

The new merged data will appear in your folder and is now ready to be used!